### ISSUE TITLE
Modify (discuss necessity for) default flags for `traverse` command such as: `--push`, `--push-untracked`

### ISSUE THREAD

**Issue by amalota** [PawelLipski/git-machete#709]
So atm both flags: `--push` and `--push-untracked` are turned on by default. In my reasoning there is no need for the `--push-untracked` flag to be turned on by default, here's why.

Let's say that we have repo with tracked branch `T` and untracked branch `U`.
We also have two flags which can turn off pushing: `--no-push` and `--no-push-untracked`.

Here are the possible scenarios:
- By default, with `--push` and `--push-untracked` we push both of them.
- Lets turn on `--no-push` flag, result -> both branches do not get pushed, flag `--push-untracked` had no impact 
- Lets turn on `--no-push-untracked` flag, result -> tracked branch `T` will be pushed, and again flag `--push-untracked` had no impact

-> to sum up, when `--push` flag is turned on (and it is by default), value of flag `--push-untracked` doesn't matter
which makes it useless.

@PawelLipski I am not 100% regarding this logic flow so please review if it makes sense. If it does then i will start thinking about modifying the flags so that there are no unnecessary flags and each push/no push scenario is possible.

**Comment by amalota** [comment]
@PawelLipski How i see it.

What we want to achieve:
- push all branches -> `--push` (default) (config key per https://github.com/VirtusLab/git-machete/issues/645)
- push only tracked branches -> `--no-push-untracked` / `--push-only-tracked` (has precedence over `--push`)
- push only untracked branches-> `--no-push-tracked` / `--push-only-untracked` (has precedence over `--push`)
- do not push at all-> `--no-push`

**Comment by PawelLipski** [comment]
Lemme number the points so that they can be referred to more easily:

0️⃣ Sorry I haven't mentioned that earlier, see the origin of the existing flags at https://github.com/VirtusLab/git-machete/pull/144

Esp. the part about 

> Alias traverse to machete traverse --no-push.
> Selectively push via an interactive traverse --push.

explains why 4 such flags were added, esp. why `--push` and `--push-untracked` were added despite their behavior being the default.

1️⃣  Hmm I'd say that the flags

> push only untracked branches-> --no-push-tracked / --push-only-untracked (has precedence over --push)

while useful for consistency/symmetry, have little practical use ☹️ it's unlikely that someone would even want to only push untracked branches... esp. given that once they're pushed, they become tracked (so the usefulness of the flag would be very sporadic).

2️⃣ 

>  (has precedence over `--push`)

For clarity... _precedence_ in this context means that even if `--push` is present **anywhere** in the command line, then `--no-push-tracked` always overrides it? What if one runs `git machete traverse --no-push-tracked --push`?

Note that the current behavior is that flags are applied from left to right, so a later `--push` overrides an earlier `--no-push-tracked`. In other words, no flag has a specific precedence over any other, it's all down to flag order.

**Comment by amalota** [comment]
0️⃣ I understand now that this flag is useful but i still don't see a reason why it has to be turned on by default. Let's examine the possible scenarios (this time with the knowledge that flag order matters :D) in the order from the most used to the least used.

1. push all branches -> `traverse`
2. do not push at all-> `traverse --no-push`
3. push only tracked branches -> `traverse --no-push-untracked`
4. push only untracked branches-> `traverse --no-push --push-untracked`

I think these are all the possible scenarios. In no case was the default true value for flag `--push-untracked` used, i would delete it from being default.

Also regarding no. 4, we have to provide 2 flags in order to only push untracked which is unnecessary when we can instead have flag  `--no-push-tracked`. Not only would we be using the expected  mechanism of turning off branches from being pushed as we already introduced with flag `--no-push-untracked` but also make it easier by using 1 flag instead of 2.

To sum up, i would replace the flag `--push-untracked` with `--no-push-tracked` flag instead.
Result:
1. push all branches -> `traverse`
2. do not push at all-> `traverse --no-push`
3. push only tracked branches -> `traverse --no-push-untracked`
4. push only untracked branches-> `traverse --no-push-tracked` **(changed)**

1️⃣ This is already possible by the flag `--push-untracked`. e.g: `traverse --no-push --push-untracked` -> functionality stays the same, just different flags

2️⃣ Okay, first of all i had no idea that flag order defines precedence, i just looked in the documentation and couldn't find it there as well (I made a separate PR for this -> #711 )

@PawelLipski wdyt?

**Comment by PawelLipski** [comment]
0️⃣ 

> To sum up, i would replace the flag --push-untracked with --no-push-tracked flag instead.

There's a major problem with removing `--push-untracked` flag (other than breaking compatibility and hence going v4.0.0, but this could be done once we drop support for Python 3.6 one day). See the original scenario from #144, as requested by `@asford`:

> Alias `traverse` to `machete traverse --no-push`.
> Selectively push via an interactive `traverse --push`.

Currently that's also possible in `--(no-)push-untracked` flavor:

> Alias `traverse` to `machete traverse --no-push-untracked`.
> Selectively push via an interactive `traverse --push-untracked`.

After removal of `--push-untracked` flag, this would need to be done as:

> Alias traverse to `machete traverse --no-push-untracked`.
> Selectively push via an interactive `traverse --push`.

which is going to be asymmetric and a bit awkward ☹️ 

If it comes to `--no-push-tracked`... this is so niche that I wouldn't bother adding that until someone actually requests it 😅 (as happened with the other flags and `@asford`).

1️⃣ Right, indeed it's now possible on the off chance someone needs it.

2️⃣ gr8, let's get that documented (although this behavior is generally pretty expected in well-designed CLIs, including git itself).

**Comment by amalota** [comment]
@PawelLipski I understand the problem with replacing the flag, but can we agree that the default value for `--push-untracked` can be removed since it has no impact?

**Comment by PawelLipski** [comment]
>  the default value for --push-untracked can be removed since it has no impact?

Hmm as in, set `git_machete.options.CommandLineOptions.opt_push_untracked` to `False` instead of `True` in `git_machete/options.py:33`? or in what sense do you mean a _removal of default value_ specifically in code?

**Comment by amalota** [comment]
> > the default value for --push-untracked can be removed since it has no impact?
> 
> Hmm as in, set `git_machete.options.CommandLineOptions.opt_push_untracked` to `False` instead of `True` in `git_machete/options.py:33`? or in what sense do you mean a _removal of default value_ specifically in code?

@PawelLipski First one -> set `git_machete.options.CommandLineOptions.opt_push_untracked` to `False`

**Comment by PawelLipski** [comment]
> set git_machete.options.CommandLineOptions.opt_push_untracked to False

Huh but given the current implementation, this will drastically change the default behavior of `traverse`, won't it? 🤔 as in, untracked braches won't be pushed by default (?).

Note that the other property of `CommandLineOptions` is `opt_push_tracked`, not `opt_push`. Passing `--(no-)push` sets both `opt_push_tracked` and `opt_push_untracked` to `True` (or `False`)

**Comment by PawelLipski** [comment]
Anyway, pls open a PoC PR and we can see there, might be easier then to discuss on an issue

**Comment by amalota** [comment]
> Anyway, pls open a PoC PR and we can see there, might be easier then to discuss on an issue

@PawelLipski Here's the PR with the changes that i wanted to make -> https://github.com/VirtusLab/git-machete/pull/721

**Comment by PawelLipski** [comment]
Oh okay, that's perfectly non-controversial :)
